crates/ruff_linter/src/rules/ruff/rules/sort_dunder_all.rs

```diff
@@ -0,0 +1,495 @@
+            if last_item_line_offset == locator.line_end(self.end()) {
+                (false, last_item.end())
+            } else {
+                (true, last_item_line_offset)
+            }
+        };
+        let mut prelude = locator
+            .slice(TextRange::new(self.start(), prelude_end))
+            .to_string();
+        let postlude = locator.slice(TextRange::new(postlude_start, self.end()));
+
+        let joined_items = if self.multiline {
+            let indentation = stylist.indentation();
+            let newline = stylist.line_ending().as_str();
+            prelude = format!("{}{}", prelude.trim_end(), newline);
+            join_multiline_dunder_all_items(
+                &sorted_items,
+                locator,
+                parent,
+                indentation,
+                newline,
+                needs_trailing_comma,
+            )
+        } else {
+            Some(join_singleline_dunder_all_items(&sorted_items, locator))
+        };
+
+        let new_dunder_all = joined_items.map(|items| format!("{prelude}{items}{postlude}"));
+        SortedDunderAll::Sorted(new_dunder_all)
+    }
+}
+
+impl Ranged for DunderAllValue {
+    fn range(&self) -> TextRange {
+        self.range
+    }
+}
+
+#[derive(Debug)]
+enum SortedDunderAll {
+    AlreadySorted,
+    Sorted(Option<String>),
+}
+
+fn collect_dunder_all_lines(range: TextRange, locator: &Locator) -> Option<Vec<DunderAllLine>> {
+    // Collect data on each line of `__all__`.
+    // Return `None` if `__all__` appears to be invalid,
+    // or if it's an edge case we don't care about.
+    let mut parentheses_open = false;
+    let mut lines = vec![];
+    let mut items_in_line = vec![];
+    let mut comment_in_line = None;
+    for pair in lexer::lex(locator.slice(range).trim(), Mode::Expression) {
+        let (tok, subrange) = pair.ok()?;
+        match tok {
+            Tok::Lpar | Tok::Lsqb => {
+                if parentheses_open {
+                    return None;
+                }
+                parentheses_open = true;
+            }
+            Tok::Rpar | Tok::Rsqb | Tok::Newline => {
+                if !(items_in_line.is_empty() && comment_in_line.is_none()) {
+                    lines.push(DunderAllLine::new(
+                        &items_in_line,
+                        comment_in_line,
+                        range.start(),
+                    ));
+                }
+                break;
+            }
+            Tok::NonLogicalNewline => {
+                if !(items_in_line.is_empty() && comment_in_line.is_none()) {
+                    lines.push(DunderAllLine::new(
+                        &items_in_line,
+                        comment_in_line,
+                        range.start(),
+                    ));
+                    items_in_line.clear();
+                    comment_in_line = None;
+                }
+            }
+            Tok::Comment(_) => comment_in_line = Some(subrange),
```

Comment:
Can you walk me through how the comment-handling works here? How are own-line comments handled? (It looks like `comment_in_line` is for end-of-line comments, like `"foo" # bar`.) What happens if there are multiple own-lines comments between items?

Reply:
Oh, a `DunderAllLine` can be _just_ a comment, with no item.

Reply:
Yeah, at this stage we're not trying to decide which elements "own" which comments, we're just trying to determine "on line 0 there's a comment"; "on line 1 there's two items and a comment"; "one line 2 there's an item and no comment", etc.

Later, when we pass the list of `DunderAllLine`s to `collect_dunder_all_items()`, we try to determine which elements in `__all__` "own" which comments (meaning the comments should move with the element when the element moves as part of being sorted).

Reply:
For clarity, I'm wondering if we should use an enum to represent "line with items, and optionally an end-of-line comment" vs. "own-line comment"? It took me a bit of time to get there in understanding that the struct was being repurposed in this way.

Reply:
> For clarity, I'm wondering if we should use an enum to represent "line with items, and optionally an end-of-line comment" vs. "own-line comment"? It took me a bit of time to get there in understanding that the struct was being repurposed in this way.

That's a really great suggestion -- tried it out, and it makes the code much clearer. Will push shortly.